[ENH] OWTSNE: Offload computation to separate thread

pavlin-policar opened the pull request:
##### Issue
The t-SNE widget previously performed all computation inside the main thread, while simulating a thread i.e. it would pause every once in a while and handle user action. For larger data sets, this meant the widget would freeze for a while until the widget received control again.

##### Description of changes

The widget now offloads all of the computation into a separate thread, and the widget is only responsible for preparing a `Task` specification, which is then handled by the other thread.

Other changes include:
- The widget now remembers partial results better, i.e. changing the exaggeration parameter doesn't recompute all the nearest neighbours again.
- More descriptive status message and the progress bar now takes into account all the steps of the procedure.
- Since t-SNE is deterministic, once the optimization is complete, pressing "Start" does nothing now, because the results will invariably the same.
- The PCA slider now changes its maximum value based on the number of features in the data.
- Add Info thing to indicate that the settings have been changed and the user should rerun the optimization.
- I've made the t-SNE API a bit more granular so that it's possible to gain access to the affinities (nearest neighbors). In principle, this is a bad idea, because this means the widget is tied to my `openTSNE` implementation, but I think it's justified because 1. in case we ever opt for a different implementation, removing this from the widget requires little effort and 2. finding the nearest neighbors can take a very long time (up to half of all the computation time for larger data sets) therefore the widget should be smart enough to cache results wherever possible.
- The widget now replots fairly frequently again. I did this because it seems that doing the work in a separate thread and drawing in the main one play well together, and the speed difference between updating often and updating rarely is small.

##### Includes
- [X] Code changes
- [X] Tests
- [ ] Documentation

VesnaT commented on Orange/widgets/unsupervised/owtsne.py:
```diff
@@ -1,60 +1,247 @@
+                    )
+
+                if state.is_interruption_requested():
+                    raise InterruptRequested()
```
Why raising an exception here and excepting it a few lines below? Maybe you could just return the `task`.

pavlin-policar replied:
Removed the exception altogether.

VesnaT commented on Orange/widgets/unsupervised/owtsne.py:
```diff
@@ -165,79 +342,148 @@ def _add_controls_start_box(self):
-
-    def resume(self):
-        self.__set_update_loop(self.tsne_iterator)
+            self.run_button.setText("Stop")
```
I think this line is not needed, since the text is already set in `start()`.

VesnaT commented:
Maybe you could write some tests for the TSNERunner. Each its method is easily testable. And it would also be good to check that the embedding is not modified inplace.

VesnaT commented on Orange/widgets/unsupervised/owtsne.py:
```diff
@@ -248,154 +431,159 @@ def set_data(self, data: Table):
+        # we double check that they are aliases
+        if task.pca_projection is not None:
+            self.__ensure_task_same_for_pca(task)
+            assert task.pca_projection is self.pca_projection
```
Since you already have a method with assert statements (self.__ensure_task_same_for_pca()), why not putting this line there? You'd only have to make sure that the method is called after setting `self.pca_projection` in `self.on_partial_result()`.

VesnaT commented on Orange/widgets/unsupervised/owtsne.py:
```diff
@@ -248,154 +431,159 @@ def set_data(self, data: Table):
+            assert task.pca_projection is self.pca_projection
+        if task.initialization is not None:
+            self.__ensure_task_same_for_initialization(task)
+            assert task.initialization is self.initialization
```
Same as above.

VesnaT commented on Orange/widgets/unsupervised/owtsne.py:
```diff
@@ -248,154 +431,159 @@ def set_data(self, data: Table):
+            self.__ensure_task_same_for_initialization(task)
+            assert task.initialization is self.initialization
+        if task.affinities is not None:
+            assert task.affinities is self.affinities
```
Same as above.

VesnaT commented on Orange/widgets/unsupervised/owtsne.py:
```diff
@@ -248,154 +431,159 @@ def set_data(self, data: Table):
+            assert task.affinities is self.affinities
+        if task.tsne_embedding is not None:
+            self.__ensure_task_same_for_embedding(task)
+            assert task.tsne_embedding is self.tsne_embedding
```
Same as above.

VesnaT commented on Orange/widgets/unsupervised/owtsne.py:
```diff
@@ -1,60 +1,235 @@
+                return
+
+        # Run regular optimization phase
+        while task.iterations_done < total_iterations_needed:
```
This code is duplicated (with the exception of some parameters). Consider writing a function and calling it twice with different parameters.

pavlin-policar replied:
Done.

VesnaT commented on Orange/widgets/unsupervised/owtsne.py:
```diff
@@ -1,60 +1,235 @@
+        if task.affinities is None:
+            job_queue.append(
+                (partial(cls.compute_affinities, task, state), weights["aff"])
+            )
```
How about something like this instead?
```
tasks = [(task.pca_projection, cls.compute_pca, weights["pca"]),
               (task.initialization, cls.compute_initialization, weights["init"]),
               (task.affinities, cls.compute_affinities, weights["aff"])]
job_queue = [(partial(y, task, state), w) for x, y, w in tasks if x]

```

pavlin-policar replied:
This is definitely shorter and much nicer to look at. However, this now seems a little complicated for what is basically several if statements and it takes a couple of seconds to see what's going on. May I propose a very straightforward alternative:

```python
        job_queue = []
        # Add the tasks that still need to be run to the job queue
        if task.pca_projection is None:
            job_queue.append((cls.compute_pca, weights["pca"]))

        if task.initialization is None:
            job_queue.append((cls.compute_initialization, weights["init"]))

        if task.affinities is None:
            job_queue.append((cls.compute_affinities, weights["aff"]))

        total_iterations = task.tsne.early_exaggeration_iter + task.tsne.n_iter
        if task.tsne_embedding is None or task.iterations_done < total_iterations:
            job_queue.append((cls.compute_tsne, weights["tsne"]))

        job_queue = [(partial(f, task, state), w) for f, w in all_jobs]
```

I do like your version, but I think this one will take a couple of seconds less to understand when one of us inevitably returns to this code.

VesnaT commented on Orange/widgets/unsupervised/owtsne.py:
```diff
@@ -164,79 +326,100 @@ def _add_controls_start_box(self):
+            )
+            self.controls.pca_components.setValue(
+                min(_DEFAULT_PCA_COMPONENTS, n_attrs)
+            )
```
This prevents setting the value of `pca_components` to the right value (the one saved in settings).

pavlin-policar replied:
Fixed.

VesnaT commented on Orange/widgets/unsupervised/owtsne.py:
```diff
@@ -248,154 +431,159 @@ def set_data(self, data: Table):
-        else:
-            perplexity = self.perplexity
+        # Cancel current running task
+        self.cancel()
```
Is this needed? The task is canceled when self.start() is invoked.

pavlin-policar replied:
Yes, it's needed: below this statement is an early exit condition if `data is None`. So if a job is running, and we remove the input signal (`data = None`), the running task won't be cancelled without this line.

VesnaT replied:
It seems that the widget is still running the task after removing the input signal despite that line.

VesnaT commented on Orange/widgets/unsupervised/owtsne.py:
```diff
@@ -164,79 +326,100 @@ def _add_controls_start_box(self):
+
             # PCA doesn't support normalization on sparse data, as this would
             # require centering and normalizing the matrix
             self.normalize_cbx.setDisabled(data.is_sparse())
```
In know this is an old code but still. It should be placed in `self.enable_controls()`. Besides, in this place `data` may differ from `self.data`, because the later could be set to None in `self.check_data()`. The same goes to the code above.

pavlin-policar replied:
That's neat, I didn't know this function existed.

VesnaT commented on Orange/widgets/unsupervised/owtsne.py:
```diff
@@ -248,154 +431,159 @@ def set_data(self, data: Table):
+        # When the data is invalid, it is set to `None` and an error is set,
+        # therefore it would be erroneous to clear the error here
+        if self.data is not None:
+            self.Error.clear()
```
Is is possible to have any error messages show at this point?

pavlin-policar replied:
Yes, you're right. This remained here from earlier when we were catching general errors in t-SNE optimization. We later agreed that it was better to crash so the error could be reported and not hidden away in the red error box thing.

pavlin-policar commented on Orange/widgets/unsupervised/tests/test_owtsne.py:
```diff
@@ -24,66 +37,70 @@ def setUpClass(cls):
+            if str(e) != "stop called on unstarted patcher":
+                raise e
 
     def restore_mocked_functions(self):
```
Kind of. It's called once now. Also, the way I'm using the patchers is pretty unconventional (I haven't seen this type of usage anywhere), and it's turned on by default, so t-SNE isn't actually run while testing the widget. But, if we want to turn this off, and _actually_ run t-SNE, we can do that with this function, without having to scratch our heads on how to do that.

VesnaT commented:
The following still happens:
- OWFile (Iris) -> OWtSNE (Check Preserve global structure -> Info message appears) -> OWFile (Reload file) -> OWtSNE (Info message has disappeared)
- OWFile (Iris) -> OWtSNE
				-> OWDataTable (Select 10 rows) -> OWtSNE (the same widget as above (input subset)) -> OWtSNE (Check Preserve global structure) -> OWDataTable (the same widget as above - reselect 10 rows) -> Crash

pavlin-policar commented:
@VesnaT I fixed the two issues you found above. I've fixed them in a very simple and hacky way. I tried to find an elegant solution, but any of these would require touching the base widget and then fixing the other widgets.
